active/0000-custom-validation-decorators.md

```diff
@@ -0,0 +1,242 @@
+
+- Using a locally defined validator
+```bicep
+func myLocalValidator(arg string, extraArg int) object => { kind: 'success' }
+
+@myLocalValidator(10)
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "__bicep",
+      "members": {
+        "myLocalValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            },
+            {
+              "name": "extraArg",
+              "type": "int"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "__bicep",
+        "name": "myLocalValidator",
+        "additionalArguments": [10]
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+- Using an imported validator
+```bicep
+import { anImportedValidator } from 'shared.bicep'
+
+@anImportedValidator()
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "__bicep",
+      "members": {
+        "anImportedValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "__bicep",
+        "name": "anImportedValidator"
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+- Using a wildcard-imported validator
+```bicep
+import * as shared from 'shared.bicep'
+
+@shared.anImportedValidator()
+param p string
+```
+
+```json
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "languageVersion": "2.0",
+  "functions": [
+    {
+      "namespace": "_1",
+      "members": {
+        "anImportedValidator": {
+          "parameters": [
+            {
+              "name": "arg",
+              "type": "string"
+            }
+          ],
+          "output": {
+            "type": "object",
+            "value": {
+              "kind": "success"
+            }
+          }
+        }
+      }
+    }
+  ],
+  "parameters": {
+    "p": {
+      "type": "string",
+      "userDefinedConstraint": {
+        "namespace": "_1",
+        "name": "anImportedValidator"
+      }
+    }
+  },
+  "resources": {}
+}
+```
+
+## Drawbacks
+
```

Comment:
just wondering, would a `throw()` function potentially help with this? When evaluated, the function would throw a `TemplateValidationException` in the backend.